File: docs/validation/python/app/scenarios/s07_opcua.py
import tempfile
from pathlib import Path
from machine_config import MachineConfigReader, MachineConfigWriter

"""
S-07: OPCUA configuration round-trip: read, write, re-read, verify OPCUA data preserved
ID:          S-07
Title:       Read, write, and re-read OPCUA fixture
Category:    happy-path
Layer:       reader + writer (OPCUA path)
Precondition: fixtures/reference_config_opcua.h5
Action:      Read → write to temp → read temp
Expected:    opcua.client.server_url unchanged
             opcua.client.session_timeout unchanged
             opcua.triggers_enabled unchanged
             All trigger names preserved
             "Chamber Oxygen Level" trigger: signal, subsystem, rule_enabled,
             start_value, stop_value all unchanged
Rationale:   OPCUA is an optional complex subgraph. Silent data loss in triggers
             would not be caught by scalar field checks.
"""

def run(fixtures_dir: str, real_dir: str) -> tuple[bool, str]:
    path = Path(fixtures_dir) / "reference_config_opcua.h5"
    if not path.exists():
        return False, f"OPCUA fixture not found: {path}"

    cfg = MachineConfigReader(path).parse()
    if cfg.opcua is None:
        return False, "opcua is None after reading OPCUA fixture"

    orig_url = cfg.opcua.client.server_url
    orig_timeout = cfg.opcua.client.session_timeout
    orig_triggers_enabled = cfg.opcua.triggers_enabled
    orig_trigger_names = set(cfg.opcua.triggers)

    with tempfile.NamedTemporaryFile(suffix=".h5", delete=False) as f:
        tmp = f.name

    MachineConfigWriter(cfg).write(tmp)
    rb = MachineConfigReader(tmp).parse()

    if rb.opcua is None:
        return False, "opcua is None after roundtrip"
    if rb.opcua.client.server_url != orig_url:
        return False, f"server_url changed: {orig_url!r} → {rb.opcua.client.server_url!r}"
    if rb.opcua.client.session_timeout != orig_timeout:
        return False, f"session_timeout changed: {orig_timeout} → {rb.opcua.client.session_timeout}"
    if rb.opcua.triggers_enabled != orig_triggers_enabled:
        return False, f"triggers_enabled changed: {orig_triggers_enabled} → {rb.opcua.triggers_enabled}"

    rb_names = set(rb.opcua.triggers)
    if rb_names != orig_trigger_names:
        return False, f"trigger names changed: missing={orig_trigger_names - rb_names}"

    co = "Chamber Oxygen Level"
    if co in cfg.opcua.triggers:
        ot, rt = cfg.opcua.triggers[co], rb.opcua.triggers[co]
        if ot.signal != rt.signal or ot.subsystem != rt.subsystem:
            return False, f"'{co}' signal/subsystem changed"

    return True, f"OPCUA roundtrip OK: {len(orig_trigger_names)} triggers, url={orig_url!r}"

File: docs/validation/rust/app/src/scenarios/s07_opcua.rs
// S-07: OPCUA config roundtrip
//
// ID:          S-07
// Precondition: fixtures/reference_config_opcua.h5
// Action:      Read → write to temp → read temp
// Expected:    server_url, session_timeout, triggers_enabled, trigger names,
//              and "Chamber Oxygen Level" signal/subsystem all unchanged.

use machine_config::{MachineConfigReader, MachineConfigWriter};
use std::collections::HashSet;
use std::path::Path;

pub fn run(fixtures_dir: &Path, _real_dir: &Path) -> (bool, String) {
    let path = fixtures_dir.join("reference_config_opcua.h5");
    if !path.exists() {
        return (false, format!("OPCUA fixture not found: {}", path.display()));
    }

    let cfg = match MachineConfigReader::open(&path).and_then(|r| r.parse()) {
        Ok(c) => c,
        Err(e) => return (false, format!("read failed: {e}")),
    };
    let opcua = match &cfg.opcua {
        Some(o) => o,
        None => return (false, "opcua is None after reading OPCUA fixture".to_string()),
    };

    let orig_url = opcua.client.server_url.clone();
    let orig_timeout = opcua.client.session_timeout;
    let orig_triggers_enabled = opcua.triggers_enabled;
    let orig_trigger_names: HashSet<String> = opcua.triggers.keys().cloned().collect();

    let tmp = match tempfile::Builder::new().suffix(".h5").tempfile() {
        Ok(t) => t,
        Err(e) => return (false, format!("tempfile creation failed: {e}")),
    };
    if let Err(e) = MachineConfigWriter::new(&cfg).write(tmp.path()) {
        return (false, format!("write failed: {e}"));
    }
    let rb = match MachineConfigReader::open(tmp.path()).and_then(|r| r.parse()) {
        Ok(c) => c,
        Err(e) => return (false, format!("readback failed: {e}")),
    };
    let rb_opcua = match &rb.opcua {
        Some(o) => o,
        None => return (false, "opcua is None after roundtrip".to_string()),
    };

    if rb_opcua.client.server_url != orig_url {
        return (false, format!("server_url changed: {orig_url:?} -> {:?}", rb_opcua.client.server_url));
    }
    if rb_opcua.client.session_timeout != orig_timeout {
        return (
            false,
            format!("session_timeout changed: {orig_timeout} -> {}", rb_opcua.client.session_timeout),
        );
    }
    if rb_opcua.triggers_enabled != orig_triggers_enabled {
        return (
            false,
            format!("triggers_enabled changed: {orig_triggers_enabled:?} -> {:?}", rb_opcua.triggers_enabled),
        );
    }

    let rb_names: HashSet<String> = rb_opcua.triggers.keys().cloned().collect();
    if rb_names != orig_trigger_names {
        let missing: Vec<_> = orig_trigger_names.difference(&rb_names).collect();
        return (false, format!("trigger names changed: missing={missing:?}"));
    }

    let co = "Chamber Oxygen Level";
    if let (Some(ot), Some(rt)) = (opcua.triggers.get(co), rb_opcua.triggers.get(co)) {
        if ot.signal != rt.signal || ot.subsystem != rt.subsystem {
            return (false, format!("'{co}' signal/subsystem changed"));
        }
    }

    (
        true,
        format!("OPCUA roundtrip OK: {} triggers, url={orig_url:?}", orig_trigger_names.len()),
    )
}
